torch/distributed/device_mesh.py

```diff
@@ -570,100 +319,183 @@ def _init_process_groups(
+
+        DeviceMesh can also be used as a context manager when using together with DTensor APIs.
+
+        .. note::
+            DeviceMesh follows SPMD programming model, which means the same PyTorch Python program
+            is running on all processes/ranks in the cluster. Therefore, users need to make sure the
+            `mesh` array (which describes the layout of devices) should be identical across all ranks.
+            Inconsistent `mesh` will lead to silent hang.
+
+        Args:
+            device_type (str): The device type of the mesh. Currently supports: "cpu", "cuda/cuda-like".
+            mesh (ndarray): A multi-dimensional array or an integer tensor describing the layout
+                of devices, where the IDs are global IDs of the default process group.
+
+        Returns:
+            DeviceMesh: A :class:`DeviceMesh` object representing the device layout.
+
+        The following program runs on each process/rank in an SPMD manner. In this example, we have 2
+        hosts with 4 GPUs each.
+        A reduction over the first dimension of mesh will reduce across
+        columns (0, 4), .. and (3, 7), a reduction over the second dimension
+        of mesh reduces across rows (0, 1, 2, 3) and (4, 5, 6, 7).
+
+        Example::
+
+            >>> # xdoctest: +SKIP("no rank")
+            >>> from torch.distributed.device_mesh import DeviceMesh
+            >>>
+            >>> # Initialize device mesh as (2, 4) to represent the topology
+            >>> # of cross-host(dim 0), and within-host (dim 1).
+            >>> mesh = DeviceMesh(device_type="cuda", mesh=[[0, 1, 2, 3],[4, 5, 6, 7]])
+        """
+
+        storage: DeviceMeshStorage
+        layouts: list[Layout]
+        mesh_dim_names: Optional[tuple[str, ...]]
+
+        def __init__(
+            self,
+            device_type: str,
+            mesh: Union[torch.Tensor, "ArrayLike"],
+            *,
+            mesh_dim_names: Optional[tuple[str, ...]] = None,
+            backend_override: Optional[
+                tuple[tuple[Optional[str], Optional[C10dBackend.Options]], ...]
+            ] = None,
+            _init_backend: bool = True,
+        ) -> None:
+            if not isinstance(mesh, torch.Tensor):
+                mesh = torch.tensor(mesh, device="cpu", dtype=torch.int)
+            elif not mesh.is_cpu:
+                raise ValueError("...")
+            mesh_dim_names = tuple(mesh_dim_names) if mesh_dim_names else None
+            if backend_override is None:
+                backend_override = ((None, None),) * mesh.ndim
+
+            self.storage = DeviceMeshStorage(device_type, _init_backend=_init_backend)
+            self.layouts = rank_tensor_to_layouts(mesh)
+            self.mesh_dim_names = mesh_dim_names
+
+            for i, layout in enumerate(self.layouts):
+                backend_override_ = backend_override[i]
+                global_override = _mesh_resources.mesh_dim_group_options.get(i, (None, None))
```

Comment:
can we put mesh_dim_group_options to the DeviceMeshStorage class as well or you plan to do in follow up PRs?

Reply:
I'd argue that `_mesh_resources.mesh_dim_group_options` is a "legacy" feature (now that we've added `backend_override` to most functions/methods), thus the way forward would be to deprecate/remove it. I'm keeping it for now though...

Reply:
Make sense!